docs/developer/testing-in-fedbiomed.md

```diff
@@ -0,0 +1,163 @@
+
+### 10. Avoid mocking the methods of the class being tested
+- Please avoid mocking public/private methods of the class being tested.
+- You can mock in some cases if you think that you have to mock it to reduce complexity. However, the need for mocking can be due to some desing issues. Therefore, please make sure that the class being tested or the test case is designed well. 
+
+### 11. Leveling Mocks
+- The classes being tested may require instantiating other instances that contain methods that need to be mocked, such as a database connection or file read/write operation. In such scenarios, instead of mocking the built-in read/write methods, mock the higher-level instance that is used by the class being tested.
+
+### 12. Private methods of a class
+- Write separate tests for private methods if it has complicated state preparation.
+- For some cases, if a private method is not complex and easy to cover every line of code, you can test them through public method. However, there is no need to test for the second time through another public method if it is already tested once. 
+
+### 13. Generic Classes that Accept Different Data Types (e.g., Dict)
+- When testing generic classes that can accept different data types, such as dictionaries containing different types of data, consider the following:
+    - Test different data types if the behavior of the class depends on the specific data types.
+    - If the class uses another class to manage different data types, and that class has been tested, there is no need to test all possible combinations of data types.
+
+### 14. Testing Different Methods that Accept Different Data Types for a Single Argument
+- Test each data type separately (e.g., float, int) to cover all possible scenarios.
+- IMPORTANT: If the behavior of the method is consistent across different data types, there is no need to test all combinations.
+
+### 15. Having complex unit test cases in the end
+- If unit tests require too much mocking and complexity, it is better to reconsider the design chosen for the class being tested. Following SOLID principles while designing a class or a module can help.
+    - Single Responsibility Principle
+        - A class should have only one reason to change.
+    - Open-Closed Principle
+        - Classes should be open for extension and closed to modification. Adding new functionality should not require modifying the existing code for the class, as it introduces potential bugs.
+    - Liskov Substitution Principle
+        - Subclasses should be substitutable for their base classes.
+    - Interface Segregation Principle
+        - Specific interfaces are better than one general-purpose interface.
+    - Dependency Inversion Principle
+        - Classes should depend on interfaces or abstract classes instead of concrete classes and functions.
+
```

Comment:
I would suggest adding one condition related to the fact that when a bug is spotted, it is good practice to create a specific unit test that reproduces the bug, even if it seems like it would break one or more of the rules above. 
At least, that's my opinion and we can discuss it.